### PR TITLE
Revamp CONTRIBUTING.md [SKIP CI]

### LINKED ISSUE #1399: Contributing.md  should be updated to represent local testbed need correctly
https://github.com/vmware/docker-volume-vsphere/blob/master/CONTRIBUTING.md#build-test-rinse-repeat

see below
```
Test environment typically consist of 1 ESX and 2 guest VMs running inside of the ESX. We also support 1 ESX and 1 guest VM. We require ESX 6.0 and later, and a Linux VM running Docker 1.10+ enabled for plain text TCP connection, i.e. Docker Daemon running with "-H tcp://0.0.0.0:2375 -H unix:///var/run/docker.sock" options. Note that both tcp: and unix: need to be present Please check "Configuring and running Docker" (https://docs.docker.com/engine/admin/configuring/) page on how to configure this - also there is a github link at the bottom, for systemd config files.
```

With the current automated tests we have added some pre-requisite for the local testbed to be configured.

- swarm tests needs swarm cluster to be pre-configured (which requires 3 vms in the test bed)
- some tests requires two vms to be present

/CC @msterin

### PR DESCRIPTION
**Pull request by shuklanirdesh82**
Fixes #1399 (Revamping CONTRIBUTING.md)

Refreshing test setup requirement and bring contributing.md to current.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -177,7 +176,7 @@ If you want to user another DockerHub repository you need to set `DOCKER_HUB_REP
   - `GOVC_URL` same as `ESX IP`
   - `GOVC_USERNAME` & `GOVC_PASSWORD`: user credentials logging in to `ESX IP`
 
-- You **need** to set following environment variables to run swarm cluster related tests.
+- You **need** to set following environment variables to run swarm cluster related tests. You **need** to configure swarm cluster inorder to run swarm related testcase otherwise there will be a test failure.
```
Comment (shaominchen) [inline CONTRIBUTING.md]: Nit: inorder => in order

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -196,6 +195,7 @@ or
 export ESX=10.20.105.54
 export VM1=10.20.105.121
 export VM2=10.20.104.210
+export VM2=10.20.104.241
```
Comment (shaominchen) [inline CONTRIBUTING.md]: VM2 => VM3

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -146,9 +146,9 @@ of binaries in ./build directory.
 
 In order to test locally, you'd need a test setup. Local test setup automation is planned but but not done yet, so currently the environment has to be set up manually.
```
Comment (msterin) [inline CONTRIBUTING.md]: Please add a subheader for "Local Test bed setuop" , currently it's mixed up with other stuff

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -166,8 +166,7 @@ Makefile targets. There targets rely on environment variables to point to the
 correct environment.
 
 Environment variables:
-- You **need** to set ESX and either VM_IP (in which case we'll use 1 VM) or
-both VM1 and VM2 environment variables
+- You **need** to set ESX and VM1, VM2 and VM3 environment variables
```
Comment (msterin) [inline CONTRIBUTING.md]: and if I don't , will the test tell me that ? It *has* to IMO, it is super cheap to check for .At the very least, tell how the failure will look like if the vars are not set up

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -146,9 +146,9 @@ of binaries in ./build directory.
 
 In order to test locally, you'd need a test setup. Local test setup automation is planned but but not done yet, so currently the environment has to be set up manually.
 
-Test environment  typically consist of 1 ESX and 2  guest VMs running inside of the
-ESX. We also support 1 ESX and 1 guest VM. We require ESX 6.0 and later,
-and a Linux VM running  Docker 1.10+ enabled for  plain text TCP connection, i.e.
+Test environment  typically consist of 1 ESX and 3 guest VMs running inside of the
```
Comment (msterin) [inline CONTRIBUTING.md]: local test bed is getting more and more complicated to set up :-(
There should be a way to do basic tests on a single VM. (not related to documentation)

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -177,7 +176,7 @@ If you want to user another DockerHub repository you need to set `DOCKER_HUB_REP
   - `GOVC_URL` same as `ESX IP`
   - `GOVC_USERNAME` & `GOVC_PASSWORD`: user credentials logging in to `ESX IP`
 
-- You **need** to set following environment variables to run swarm cluster related tests.
+- You **need** to set following environment variables to run swarm cluster related tests. You **need** to configure swarm cluster in order to run swarm related testcase otherwise there will be a test failure.
```
Comment (msterin) [inline CONTRIBUTING.md]: Oh, so now I need 3 VMS **and** a swarm cluster? Or do I need to configure these VMs into a cluster? 
Please provide instructions or reference to configure swarm too.

Comment (msterin) [inline CONTRIBUTING.md]: Do I need to configure any env for SWARM ? MASTER ? SLAVE(s) ?

File: CONTRIBUTING.md (hunk 5 above)
Comment (shaominchen) [inline CONTRIBUTING.md]: Agree with Mark. Nirdesh and I discussed this and we agree it's too heavy to have the basic precheckin tests running on a swarm cluster (at least 3 nodes) environment. For internal developers, we can use nimbus to deploy a testbed with one command. But for external developers, it will be too complicated to set up the environment simply for running the precheckin test.

On the other hand, we also want to make sure our CI can capture most common issues by running all P0 and part of P1 test cases. Also we should still support running all complicated test cases locally with some optional parameters

To achieve these different goals, here's our plan:
1. Developers should be able to run test-all on a single VM. This is achieved by moving all extensive test cases (which require complicated setup) to one or more special labels
2. Developers should be able to run complicated test case by specifying some optional parameters. The assumption is that the environment should be ready for running those tests
3. CI should continue to all extensive test cases as we have already planned

Let me know if you have any comments @msterin @shuklanirdesh82 . I will file a issue to track this task.

File: CONTRIBUTING.md (hunk 3 above)
Comment (shuklanirdesh82) [inline CONTRIBUTING.md]: done!

File: CONTRIBUTING.md (hunk 4 above)
Comment (shuklanirdesh82) [inline CONTRIBUTING.md]: Yeah the error message improvement will be handled separately in the different PR. More detailed information is added.

File: CONTRIBUTING.md (hunk 6 above)
Comment (shuklanirdesh82) [inline CONTRIBUTING.md]: I have added a `TODO` information ... and will work on simplifying the `make test-all` invocation. It will be handled separately in different PR having it said keeping this PR only for documentation changes.

File: CONTRIBUTING.md (hunk 5 above)
Comment (shaominchen) [inline CONTRIBUTING.md]: Filed issue #1505

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -144,11 +77,18 @@ Use `make` or `make dockerbuild` to build.
...
+### Local Testbed setup
+
+In order to test locally, you'd need a test setup. Local testbed setup automation is planned but not done yet, so currently the environment has to be set up manually.
+
+Test environment  typically consist of 1 ESX and 3 guest VMs running inside of the ESX to run all automated tests.
 
-Test environment  typically consist of 1 ESX and 2  guest VMs running inside of the
-ESX. We also support 1 ESX and 1 guest VM. We require ESX 6.0 and later,
-and a Linux VM running  Docker 1.10+ enabled for  plain text TCP connection, i.e.
+**TODO**: As a next step, an approach will be provided to run selective test and above testbed requirement will be simplified. Basic test run will not be block when minimum testbed requirement is met i.e. 1 ESX and 1 or 2 VMs.
```
Comment (pshahzeb) [inline CONTRIBUTING.md]: will not be block  -> will not be blocked

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -144,11 +77,18 @@ Use `make` or `make dockerbuild` to build.
...
+
+**Note**: All 3 guest VMs can be resued to run swarm testcase by configuring swarm cluster (by making one node as master and registered others as worker nodes). It is not mandatory to create extra VM to cofigure swarm cluster but feel free if you would like to do so. Please make sure to set environment variables correctly as mentioned in section below.
```
Comment (pshahzeb) [inline CONTRIBUTING.md]: cofigure -> configure

Comment (ashahi1) [inline CONTRIBUTING.md]: resued -> reused

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -177,7 +116,8 @@ If you want to user another DockerHub repository you need to set `DOCKER_HUB_REP
   - `GOVC_URL` same as `ESX IP`
   - `GOVC_USERNAME` & `GOVC_PASSWORD`: user credentials logging in to `ESX IP`
 
-- You **need** to set following environment variables to run swarm cluster related tests.
+- You **need** to set following environment variables to run swarm cluster related tests. You **need** to configure swarm cluster in order to run swarm related testcase otherwise there will be a test failure. As mentioned above, you may reuse the same set of VMs here no need to create separately.
+
```
Comment (ashahi1) [inline CONTRIBUTING.md]: here no -> here, no

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -2,17 +2,29 @@
...
+  * [Setting up local testbed](#local-testbed-setup)
+  * [Capturing ESX Code Coverage](#capturing-esx-code-coverage)
 * [Bug filing guidelines](#bug-filing-guidelines)
-* [Testing and CI](#testing-and-ci)
+* [Testing on CI](#testing-on-ci-system)
+  * [Current CI setup](#ci-system)
+* [Cutting a new release guidelines](#cutting-a-new-release-guidelines)
+  * [Tagging a release](#tag-a-release)
+  * [Generating change log](#generate-the-change-log)
+  * [Publishing artifects](#publish-vdvs-managed-plugin-to-docker-store)
```
Comment (ashahi1) [inline CONTRIBUTING.md]: Please check if it is artifects or artifacts?

Comment (shuklanirdesh82) [inline CONTRIBUTING.md]: done!

File: CONTRIBUTING.md (hunk 7 above)
Comment (shuklanirdesh82) [inline CONTRIBUTING.md]: done!

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -1,18 +1,26 @@
...
 Read the [FAQ on the Wiki](https://github.com/vmware/docker-volume-vsphere/wiki#faq)
 ### Pull Requests
 * Create a fork or branch (if you can) and make your changes
-   * Branch should be suffixed with github user id: (branch name).(github user id) Example: mydevbranch.kerneltime
-   * If you want to trigger CI test runs for pushing into a branch prefix the branch with runci/ Example: runci/mylatestchange.kerneltime
-   * If you want to skip running CI test e.g. *Documentation change/Inline comment change* or when the change is in `WIP` or `PREVIEW` phase, add `[CI SKIP]` or `[SKIP CI]` to the PR title.
+   * Branch should be suffixed with github user id: `(branch name).(github user id)` Example: `mydevbranch.kerneltime`
+   * If you want to trigger CI test runs for pushing into a branch prefix the branch with `runci/` Example: `runci/mylatestchange.kerneltime`
+   * If you want to skip running CI test e.g. **Documentation change/Inline comment change** or when the change is in `WIP` or `PREVIEW` phase, add `[CI SKIP]` or `[SKIP CI]` to the PR title.
 * Each PR must be accompanied with unit/integration tests
```
Comment (shaominchen) [inline CONTRIBUTING.md]: Nit: Not your change, but I think we should say "Each PR **should** be accompanied with unit/integration tests whenever possible"

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -138,17 +67,38 @@ Copy-paste relevant snippets from log file, code. Use github markdown language f
 
 ```
 ## Typical Developer Workflow
+
+A typical workflow for a developer should be.
+
+- Create a branch, push changes and make sure tests do not break as reported
+  by the CI system.
+- When ready post a PR. This will trigger a full set of tests on ESX. After all
```
Comment (shaominchen) [inline CONTRIBUTING.md]: Nit: When ready post a PR. This will... =>When ready posting a PR, it will...

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -138,17 +67,38 @@ Copy-paste relevant snippets from log file, code. Use github markdown language f
...
 
-Test environment  typically consist of 1 ESX and 2  guest VMs running inside of the
-ESX. We also support 1 ESX and 1 guest VM. We require ESX 6.0 and later,
-and a Linux VM running  Docker 1.10+ enabled for  plain text TCP connection, i.e.
+**TODO**: As a next step, an approach will be provided to run selective test and above testbed requirement will be simplified. Basic test run will not be blocked when minimum testbed requirement is met i.e. 1 ESX and 1 or 2 VMs.
+
+**Note**: All 3 guest VMs can be reused to run swarm testcase by configuring swarm cluster (by making one node as master and registered others as worker nodes). It is not mandatory to create extra VM to configure swarm cluster but feel free if you would like to do so. Please make sure to set environment variables correctly as mentioned in section below.
+
+We require ESX 6.0 and later,
+and a Linux VM running  Docker 1.13+ enabled for  plain text TCP connection, i.e.
```
Comment (shaominchen) [inline CONTRIBUTING.md]: Nit: remove redundant spaces in this line

File: CI.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+- Each checkin into a branch on the official repo will run the full set of
+  tests.
+
+- On Pull Requests the full set of tests will be run.
+  - The logs for the build and test will be posted to the CI server.
+  - Due to security concerns the artifacts will not be published.
+    https://github.com/drone/drone/issues/1476
+
+- Each commit into the master operation will run the full set of tests
```
Comment (shuklanirdesh82) [inline CI.md]: done!

File: CI.md [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+    (and `<branch>-<build>`). Any one can pull this image and run it to
+    use the docker plugin.
+
+## CI System
+
+We are using the CI system that has been up by the CNA folks (@casualjim, @frapposelli & @mhagen-vmware).
+The CI system is based on https://drone.io/  URL for the server is  https://ci.vmware.run/
+To be able to change the integration between CI and GitHub, first become admin using the self serve portal.
+
+Behind the firewall there is a HW node running ESX that hosts 2 ESX VMs (v6.0u2 and v6.5). Each VM in turn contains a VSAN datastore as well as a VMFS datastore with 2 VMs per datastore (VMs are reused to form swarm cluster). CI testbed is also configured to have swarm cluster and there 3 VMs participating into the swarm cluster (2 VMs created on vmfs datastore and 1 VM created on vsan datastore).
```
Comment (shuklanirdesh82) [inline CI.md]: done!
